tests/python_package_test/test_dask.py

```diff
@@ -446,11 +446,29 @@ def test_classifier_pred_contrib(output, task, cluster):
                 assert len(np.unique(preds_with_contrib[:, base_value_col]) == 1)
 
 
-def test_find_random_open_port(cluster):
+def test_group_workers_by_host():
+    hosts = [f'0.0.0.{i}' for i in range(2)]
+    workers = [f'tcp://{host}:{p}' for p in range(2) for host in hosts]
+    expected = {
+        host: lgb.dask._HostWorkers(
+            default=f'tcp://{host}:0',
+            all=[f'tcp://{host}:0', f'tcp://{host}:1']
+        )
+        for host in hosts
+    }
+    host_to_workers = lgb.dask._group_workers_by_host(workers)
+    assert host_to_workers == expected
+
+
+def test_assign_open_ports_to_workers(cluster):
     with Client(cluster) as client:
-        for _ in range(5):
-            worker_address_to_port = client.run(lgb.dask._find_random_open_port)
+        workers = client.scheduler_info()['workers'].keys()
+        n_workers = len(workers)
+        host_to_workers = lgb.dask._group_workers_by_host(workers)
+        for _ in range(1_000):
```

Comment:
Probably this line was the reason of just exceeded time limit QEMU CI job at `master`. Will keep an eye on it.

Reply:
I think you're probably right!

I didn't consider it, since I've gotten into the habit of just thinking "well, sometimes setup on QEMU builds is really slow, and it's unpredictable whether that job will take 90 minutes or 120 minutes".

But I just compared the logs on latest `master` (with this PR merged) to the logs from the previous `master` build, and the Dask tests took about 22 minutes longer to run than they did previously.

I'd support a PR to drop this to `range(100)` or even `range(25)`.

**commit with this PR (https://github.com/microsoft/LightGBM/commit/5fe27d594285eeaf31d2f1ebfe29453f8c6c6ad3)**

* total runtime: 2h7m
* time for Dask tests: 58m08s
* link: https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=10646&view=logs&j=380bc479-1396-5123-5f0f-4e5f44cf253e&t=d4eba53f-d934-5d3f-b687-2f058f88f797

```text
2021-08-04T12:35:22.0970407Z ../tests/python_package_test/test_consistency.py ......                  [  9%]
2021-08-04T12:40:11.5075701Z ../tests/python_package_test/test_dask.py .............................. [ 14%]
2021-08-04T12:50:29.7834725Z ........................................................................ [ 26%]
2021-08-04T13:08:08.7403444Z ........................................................................ [ 37%]
2021-08-04T13:26:40.1951149Z ......s...............s...............s...............s................. [ 49%]
2021-08-04T13:36:51.6836691Z ..................................s..................................... [ 61%]
2021-08-04T13:38:19.0058818Z s.......................s........                                        [ 66%]
2021-08-04T13:38:19.0156018Z ../tests/python_package_test/test_dual.py s                              [ 66%]
```

**previous commit on `master` (https://github.com/microsoft/LightGBM/commit/1dbf4382fd7d8cd6ce92b53c3b84aa479aabfc19)**

* total runtime: 1h15m
* time for Dask tests: 36m22s
* link: https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=10642&view=logs&j=c2f9361f-3c13-57db-3206-cee89820d5e3&t=bf5b33f9-9072-549e-d8b1-cab79a1dd2a9

```text
2021-08-03T21:00:22.0875288Z ../tests/python_package_test/test_consistency.py ......                  [  9%]
2021-08-03T21:03:26.1542925Z ../tests/python_package_test/test_dask.py .............................. [ 14%]
2021-08-03T21:09:08.6432850Z ........................................................................ [ 26%]
2021-08-03T21:18:07.7971801Z ........................................................................ [ 37%]
2021-08-03T21:32:02.3425272Z ......s...............s...............s...............s................. [ 49%]
2021-08-03T21:38:51.5159320Z ..................................s..................................... [ 61%]
2021-08-03T21:39:48.7253040Z s.......................s........                                        [ 66%]
2021-08-03T21:39:48.7338523Z ../tests/python_package_test/test_dual.py s                              [ 66%]
```

Reply:
I just added @jameslamb's suggestion in #4501